File: programmers/level2/메뉴 리뉴얼.py
from itertools import combinations
from collections import defaultdict


def solution(orders, course):
    answer = []

    for c in course:
        combiCount = defaultdict(int)  # 각 조합별 개수를 저장하는 딕셔너리
        maxOrder = []  # 가장 큰 value를 갖는 key값 저장하는 리스트
        for order in orders:
            orderCombi = combinations(list(order), c)  # 코스의 단품 개수만큼 조합 추출
            for oc in orderCombi:
                orderString = ''.join(sorted(oc))  # 문자열 정렬해서 합침
                combiCount[orderString] += 1  # 딕셔너리에 해당 키값의 value 증가
        maxOrder = [k for k, v in combiCount.items() if
                    ((v == max(combiCount.values())) and v >= 2)]  # v가 maxValue와 같고, 2이상인 경우
        answer.extend(maxOrder)  # 리스트 다 제거해서 삽입
    answer = sorted(answer)  # 알파벳 순서로 정렬

    # 시간 초과
    #     result = []
    #     menu = []
    #     menu = set(menu)
    #     for o in orders:
    #         for i in range(len(o)):
    #             menu.add(o[i])

    #     menu = list(menu)
    #     menu.sort()
    #     dict = defaultdict(list)
    #     maxValueDict = defaultdict(int)

    #     for c in course:
    #         for combi in list(combinations(menu, c)):
    #             sum = 0
    #             for order in orders:
    #                 flag = True
    #                 for com in combi:
    #                     if com not in order: # 하나라도 포함 되지 않는 다면
    #                         flag = False
    #                         break
    #                 if flag:
    #                     sum += 1
    #             if sum > 1:
    #                 if maxValueDict[c] < sum:
    #                     maxValueDict[c] = sum
    #                     dict[c].clear()
    #                     dict[c].append(combi)
    #                 elif maxValueDict[c] == sum:
    #                     maxValueDict[c] = sum
    #                     dict[c].append(combi)

    #     for key, value in dict.items():
    #         for v in value:
    #             answer.append(''.join(v))

    #     answer.sort()

    return answer

File: programmers/level2/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    Map<String, Integer> hashMap = new HashMap<>();
    List<String> answerList = new ArrayList<>();

    public String[] solution(String[] orders, int[] course) {
        for (int i = 0; i < orders.length; i++) {
            char[] arr = orders[i].toCharArray(); // 문자열 내 문자들 변경
            Arrays.sort(arr); // 문자 정렬
            orders[i] = String.valueOf(arr); // 문자열로 변경
        }

        for (int courseLength: course) {
            for (String order: orders) { // course 수에 맞는 조합 추출
                combination("", order, courseLength); // 각 order에 맞는 조합 추출
            }

            if (!hashMap.isEmpty()) {
                List<Integer> countList = new ArrayList<>(hashMap.values()); // map의 value만 List로 변환
                int max = Collections.max(countList); // 최대값 계산

                if (max > 1) { // 중복 메뉴가 2번 이상인 경우에만
                    for (String key: hashMap.keySet()) { // map의 key값들 순회하면서 체크
                        if (hashMap.get(key) == max) {
                            answerList.add(key);
                        }
                    }
                }
            }
            hashMap.clear(); // hashmap 초기화
        }

        Collections.sort(answerList); // 알파벳순 정렬
        String[] answer = new String[answerList.size()];
        for (int i = 0; i < answer.length; i++) {
            answer[i] = answerList.get(i);
        }

        return answer;
    }

    public void combination(String order, String others, int count) {
        if (count == 0) { // 조합 계산이 다 된 경우
            // System.out.println("order = " + order);
            hashMap.put(order, hashMap.getOrDefault(order, 0) + 1); // map에 추가
            return;
        }
        // System.out.println("check = " + order + " others = " + others);

        for (int i = 0; i < others.length(); i++) {
            // System.out.println("i = " + i + " before = " + order + " : " + others + "=>  " + count);
            combination(order + others.charAt(i), others.substring(i + 1), count - 1); // 재귀로 계산 
            // System.out.println("i = " + i + " after = " + order + " : " + others + "=>  " + count);
        }
    }
}